staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/complete.go

```diff
@@ -66,17 +66,29 @@ func validateNestedValueValidationCompleteness(v *NestedValueValidation, s *Stru
 
 	allErrs := field.ErrorList{}
 
-	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath)...)
-	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"))...)
+	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath, opts)...)
+	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"), opts)...)
+
+	var additionalPropertiesSchema *Structural
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		additionalPropertiesSchema = s.AdditionalProperties.Structural
+	}
+
 	for k, vFld := range v.Properties {
 		if sFld, ok := s.Properties[k]; !ok {
-			allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			if additionalPropertiesSchema == nil || !opts.AllowValidationPropertiesWithAdditionalProperties {
+				allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			} else {
+				allErrs = append(allErrs, validateNestedValueValidationCompleteness(&vFld, additionalPropertiesSchema, sPath.Child("additionalProperties"), vPath.Child("properties").Key(k), opts)...)
```

Comment:
This changes the semantics if `AllowValidationPropertiesWithAdditionalProperties` is true (intentionally), e.g.:

```yaml
allOf:
  properties:
    a:
      pattern: .*
addionalProperties:
  type: string
```
Before this PR: was not complete
After this PR: is complete.

I wonder whether not allowing this was an oversight in the original structural schema definition. Have you checked in the KEP? In my current mental model I could be convinced that this is a thing we want to allow in CRDs.

Reply:
Relevant portion of the KEP: https://github.com/kubernetes/enhancements/blob/76f27981ace434ce9d4d01805a37bbc111992341/keps/sig-api-machinery/2335-vanilla-crd-openapi-subset-structural-schemas/README.md#L317-L338

> Structurally complete means that any specified JSON path in a schema has a specified type outside of any logical junctors (`andOf`,`anyOf`,`oneOf`,`not`; compare the guiding principle above and its example). In other words, we validate that
>
> - every object field (`properties`),
>- every array value (`items`) and
>- every string map value (`additionalProperties`)
>
>has a non-empty `type` outside of logical junctors, with the possible exceptions of
>- `x-kubernetes-int-or-string: true` is set in the empty-type schema
>- `x-kubernetes-preserve-unknown-fields: true` is set in the empty-type schema.
>
>If such a `type` is empty, we report the following structural schema validation error (via the >`NonStructuralSchema` condition messages, and via real API rejection according to the [API Compatibility Plan](#api-compatibility-plan)),

No direct mention, I think this change agrees in spirit by still following convention we can resolve a non-empty `type` for any validation rule.